cpp/src_prims/stats/histogram.h

```diff
@@ -0,0 +1,468 @@
+  auto op = [=] __device__(int binId, IdxT row, IdxT col) {
+    if (row >= nrows) return;
+#if __CUDA_ARCH__ < 700
+    atomicAdd(sbins + binId, 1);
+#else
+    auto amask = __activemask();
+    auto mask = __match_any_sync(amask, binId);
+    auto leader = __ffs(mask) - 1;
+    if (laneId() == leader) {
+      atomicAdd(sbins + binId, __popc(mask));
+    }
+#endif  // __CUDA_ARCH__
+  };
+  IdxT col = blockIdx.y;
+  histCoreOp<DataT, BinnerOp, IdxT, VecLen>(data, nrows, nbins, binner, op,
+                                            col);
+  __syncthreads();
+  auto binOffset = col * nbins;
+  for (auto i = threadIdx.x; i < nbins; i += blockDim.x) {
+    atomicAdd(bins + binOffset + i, sbins[i]);
+  }
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int VecLen>
+void smemHist(int* bins, IdxT nbins, const DataT* data, IdxT nrows, IdxT ncols,
+              BinnerOp binner, cudaStream_t stream) {
+  auto blks = computeGridDim<IdxT, VecLen>(nrows, ncols);
+  size_t smemSize = nbins * sizeof(unsigned);
+  smemHistKernel<DataT, BinnerOp, IdxT, VecLen>
+    <<<blks, ThreadsPerBlock, smemSize, stream>>>(bins, data, nrows, nbins,
+                                                  binner);
+}
+
+template <unsigned BIN_BITS>
+DI void incrementBin(unsigned* sbins, int* bins, int nbins, int binId,
+                     int col) {
+  constexpr unsigned WORD_BITS = sizeof(unsigned) * 8;
+  constexpr unsigned WORD_BINS = WORD_BITS / BIN_BITS;
+  constexpr unsigned BIN_MASK = (1 << BIN_BITS) - 1;
+  auto iword = binId / WORD_BINS;
+  auto ibin = binId % WORD_BINS;
+  auto sh = ibin * BIN_BITS;
+  auto old_word = atomicAdd(sbins + iword, unsigned(1 << sh));
+  auto new_word = old_word + unsigned(1 << sh);
+  if ((new_word >> sh & BIN_MASK) != 0) return;
+  int binOffset = col * nbins;
+  // overflow
+  atomicAdd(bins + binOffset + binId, BIN_MASK + 1);
+  for (int dbin = 1; ibin + dbin < WORD_BINS && binId + dbin < nbins; ++dbin) {
+    auto sh1 = (ibin + dbin) * BIN_BITS;
+    if ((new_word >> sh1 & BIN_MASK) == 0) {
+      // overflow
+      atomicAdd(bins + binOffset + binId + dbin, BIN_MASK);
+    } else {
+      // correction
+      atomicAdd(bins + binOffset + binId + dbin, -1);
+      break;
+    }
+  }
+}
+
+template <>
+DI void incrementBin<1>(unsigned* sbins, int* bins, int nbins, int binId,
+                        int col) {
+  constexpr unsigned WORD_BITS = 32;
+  auto iword = binId / WORD_BITS;
+  auto sh = binId % WORD_BITS;
+  auto old_word = atomicXor(sbins + iword, unsigned(1 << sh));
+  int binOffset = col * nbins;
+  if ((old_word >> sh & 1) != 0) atomicAdd(bins + binOffset + binId, 2);
+}
+
+template <typename DataT, typename BinnerOp, typename IdxT, int BIN_BITS,
+          int VecLen>
+__global__ void smemBitsHistKernel(int* bins, const DataT* data, IdxT nrows,
+                                   IdxT nbins, BinnerOp binner) {
+  extern __shared__ unsigned sbins[];
+  constexpr unsigned WORD_BITS = sizeof(unsigned) * 8;
```

Comment:
These constants get repeated, so feel free to wrap them into a type parameterized by `BIN_BITS`. This is up to you, however.

Reply:
Done.